### PR TITLE
ScriptWriter refactor

### PR DESCRIPTION
**Pull request by mwhamgenomics**
Executor script writers have been refactored, and no longer need to know beforehand how large a job array to create. Instead, the writer registers a job array, then writes/prepends the header, evaluating the job array size from the number of array commands registered. Tests have also been simplified.

### CODE REVIEW

File: egcg_core/executor/script_writers/script_writer.py [hunk 1]
```diff
@@ -23,65 +25,66 @@ def __init__(self, job_name, working_dir, job_queue, jobs=1, log_commands=True):
...
-            raise ValueError('Bad number of array jobs: %s written, %s expected' % (self.jobs_written, self.job_total))
-        script_file = open(self.script_name, 'w')
-
-        for line in self.lines:
-            script_file.write(line + '\n')
-        script_file.close()
-        self.debug('Closed ' + self.script_name)
+        with open(self.script_name, 'w') as f:
+            for line in self.lines:
+                f.write(line + '\n')
```
Comment (tcezard) [inline egcg_core/executor/script_writers/script_writer.py]: maybe something like 
`f.write('\n'.join(self.lines) + '\n')`
might be more consise

File: egcg_core/executor/script_writers/script_writer.py [hunk 2]
```diff
@@ -23,65 +25,66 @@ def __init__(self, job_name, working_dir, job_queue, jobs=1, log_commands=True):
...
             self._start_array()
             for idx, cmd in enumerate(cmds):
-                if self.log_commands:
-                    self._write_array_cmd(idx + 1, cmd, log_file=self.log_file + str(idx + 1))
-                else:
-                    self._write_array_cmd(idx + 1, cmd)
+                self._register_array_cmd(
+                    idx + 1,
+                    cmd,
+                    log_file=self.log_file + str(idx + 1) if self.log_commands else None
```
Comment (tcezard) [inline egcg_core/executor/script_writers/script_writer.py:53]: Did not know this construct existed

File: egcg_core/rest_communication.py [hunk 3]
```diff
@@ -61,6 +60,9 @@ def _req(self, method, url, quiet=False, **kwargs):
             kwargs.update(headers={'Authorization': 'Token ' + self.auth})
 
         r = requests.request(method, url, **kwargs)
+
+        kwargs.pop('auth', None)
+        kwargs.pop('headers', None)
```
Comment (tcezard) [inline egcg_core/rest_communication.py:65]: Does this belong to this PR or has it been added by mistake?

**Review by tcezard — COMMENTED** [review]
It all look good to me.

File: egcg_core/rest_communication.py (hunk 3 above)
Comment (mwhamgenomics) [reply egcg_core/rest_communication.py:65]: Not specific to script writers, but I thought I'd fix it while I had a PR open

**Pull request merged**